# Heart Disease Prediction - Machine Learning Project

## Pluto Academy AI & Machine Learning Internship

### Project Objective

The objective of this project is to build a machine learning classification system that predicts whether a person has heart disease based on health-related features.

The project includes data preprocessing, feature engineering, training multiple machine learning models, evaluating their performance, and selecting the best-performing model.

### Tools Used

- Python
- Pandas
- NumPy
- Matplotlib
- Seaborn
- Scikit-learn
- Google Colab

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Libraries imported successfully!")

Libraries imported successfully!


In [2]:
df = pd.read_csv("heart.csv")

print("Dataset loaded successfully!")
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Dataset loaded successfully!
Rows: 1025
Columns: 14


In [3]:
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0


# 1. Load and Inspect the Dataset

In [4]:
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0


In [5]:
print("Dataset Shape:", df.shape)

Dataset Shape: (1025, 14)


In [6]:
print("Column Names:")
print(df.columns.tolist())

Column Names:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target']


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


In [8]:
print("Missing Values:")
print(df.isnull().sum())

Missing Values:
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64


In [9]:
print("Duplicate Rows:", df.duplicated().sum())

Duplicate Rows: 723


In [10]:
print("Target Distribution:")
print(df['target'].value_counts())

Target Distribution:
target
1    526
0    499
Name: count, dtype: int64


In [11]:
print("\nTarget Percentage:")
print(df['target'].value_counts(normalize=True) * 100)


Target Percentage:
target
1    51.317073
0    48.682927
Name: proportion, dtype: float64


In [12]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,1025.0,54.434146,9.072290,29.0,48.0,56.0,61.0,77.0
sex,1025.0,0.695610,0.460373,0.0,0.0,1.0,1.0,1.0
cp,1025.0,0.942439,1.029641,0.0,0.0,1.0,2.0,3.0
trestbps,1025.0,131.611707,17.516718,94.0,120.0,130.0,140.0,200.0
chol,1025.0,246.000000,51.592510,126.0,211.0,240.0,275.0,564.0
fbs,1025.0,0.149268,0.356527,0.0,0.0,0.0,0.0,1.0
restecg,1025.0,0.529756,0.527878,0.0,0.0,1.0,1.0,2.0
thalach,1025.0,149.114146,23.005724,71.0,132.0,152.0,166.0,202.0
exang,1025.0,0.336585,0.472772,0.0,0.0,0.0,1.0,1.0
oldpeak,1025.0,1.071512,1.175053,0.0,0.0,0.8,1.8,6.2


## Dataset Summary

The Heart Disease dataset contains medical and health-related features that can be used to predict the presence of heart disease. The target variable is `target`, where 0 represents no heart disease and 1 represents the presence of heart disease. The dataset contains numerical features describing factors such as age, blood pressure, cholesterol, heart rate, and other clinical measurements. The data will be inspected for missing values and duplicate records before preprocessing. The cleaned data will then be divided into training and testing sets for machine learning.

In [13]:
print("Missing values:", df.isnull().sum().sum())
print("Duplicate rows:", df.duplicated().sum())

Missing values: 0
Duplicate rows: 723


In [14]:
print(df.dtypes)

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object


In [15]:
df.shape

(1025, 14)

In [16]:
df.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,0
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


In [17]:
df.duplicated().sum()

np.int64(723)

In [18]:
df['target'].value_counts()

,count
target,
1,526
0,499


# 2. Data Preprocessing

In [19]:
# Check duplicates before removal
print("Duplicates before removal:", df.duplicated().sum())

# Remove duplicate rows
df_clean = df.drop_duplicates().copy()

print("Duplicates after removal:", df_clean.duplicated().sum())
print("Shape after removing duplicates:", df_clean.shape)

Duplicates before removal: 723
Duplicates after removal: 0
Shape after removing duplicates: (302, 14)


In [20]:
print("Target distribution after removing duplicates:")
print(df_clean['target'].value_counts())

Target distribution after removing duplicates:
target
1    164
0    138
Name: count, dtype: int64


In [21]:
# Separate input features and target variable

X = df_clean.drop('target', axis=1)
y = df_clean['target']

print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (302, 13)
Target shape: (302,)


In [22]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("Training features:", X_train.shape)
print("Testing features:", X_test.shape)
print("Training target:", y_train.shape)
print("Testing target:", y_test.shape)

Training features: (241, 13)
Testing features: (61, 13)
Training target: (241,)
Testing target: (61,)


In [23]:
print("Training target distribution:")
print(y_train.value_counts(normalize=True))

print("\nTesting target distribution:")
print(y_test.value_counts(normalize=True))

Training target distribution:
target
1    0.543568
0    0.456432
Name: proportion, dtype: float64

Testing target distribution:
target
1    0.540984
0    0.459016
Name: proportion, dtype: float64


## Preprocessing Decisions

1. The dataset was checked for duplicate records and 723 duplicate rows were identified.
2. Duplicate records were removed to prevent repeated observations from affecting model training and evaluation.
3. The `target` column was separated from the input features.
4. The dataset was divided into training and testing sets using an 80/20 split.
5. Stratified sampling was used to maintain a similar target-class distribution in both training and testing datasets.
6. A fixed random state of 42 was used to make the train-test split reproducible.

In [24]:
df_clean.shape

(302, 14)

# 3. Feature Engineering and Scaling

In [25]:
from sklearn.preprocessing import StandardScaler

# Create the scaler
scaler = StandardScaler()

# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)

# Transform test data using the same scaler
X_test_scaled = scaler.transform(X_test)

print("Original training shape:", X_train.shape)
print("Scaled training shape:", X_train_scaled.shape)

print("Original testing shape:", X_test.shape)
print("Scaled testing shape:", X_test_scaled.shape)

Original training shape: (241, 13)
Scaled training shape: (241, 13)
Original testing shape: (61, 13)
Scaled testing shape: (61, 13)


In [26]:
print("First 5 rows of scaled training data:")
print(X_train_scaled[:5])

First 5 rows of scaled training data:
[[ 1.42194408 -1.44559455  0.98649276 -0.97304069  5.88290757 -0.39831375
  -1.00805864  0.4082403  -0.65271395  0.52726339 -0.69712735 -0.72271635
   1.07583844]
 [-0.46584105  0.6917569  -0.95428075  0.75650697 -0.88569606 -0.39831375
  -1.00805864 -1.10470453  1.53206469 -0.08323307 -0.69712735 -0.72271635
   1.07583844]
 [ 0.42252843  0.6917569  -0.95428075 -0.19772622 -0.58817502 -0.39831375
  -1.00805864 -0.88221265  1.53206469  1.05054606 -0.69712735  2.13260562
   1.07583844]
 [ 1.64403645  0.6917569   0.98649276  0.51794867  0.11843745 -0.39831375
  -1.00805864 -0.21473698 -0.65271395  0.8761185  -0.69712735  2.13260562
   1.07583844]
 [-1.02107196 -1.44559455  0.016106   -1.15195941 -1.62949866 -0.39831375
   0.8752137  -0.570724   -0.65271395 -0.86815708 -0.69712735 -0.72271635
  -0.54464321]]


## Feature Scaling

StandardScaler was used to standardize the numerical features. The scaler was fitted only on the training dataset and then applied to both the training and testing datasets. This prevents data leakage from the test set and ensures that features with different numerical ranges can be compared fairly by the machine learning algorithms.

# 4. Machine Learning Model Training

Three classification algorithms are trained and evaluated:

1. Logistic Regression
2. K-Nearest Neighbors (KNN)
3. Random Forest

In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

In [28]:
# Logistic Regression

logistic_model = LogisticRegression(random_state=42)

logistic_model.fit(X_train_scaled, y_train)

logistic_pred = logistic_model.predict(X_test_scaled)

print("Logistic Regression trained successfully!")

Logistic Regression trained successfully!


In [29]:
# K-Nearest Neighbors

knn_model = KNeighborsClassifier(n_neighbors=5)

knn_model.fit(X_train_scaled, y_train)

knn_pred = knn_model.predict(X_test_scaled)

print("KNN trained successfully!")

KNN trained successfully!


In [30]:
# Random Forest

rf_model = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf_model.fit(X_train_scaled, y_train)

rf_pred = rf_model.predict(X_test_scaled)

print("Random Forest trained successfully!")

Random Forest trained successfully!


In [31]:
# Calculate evaluation metrics

models = {
    "Logistic Regression": logistic_pred,
    "KNN": knn_pred,
    "Random Forest": rf_pred
}

results = []

for model_name, predictions in models.items():
    accuracy = accuracy_score(y_test, predictions)
    precision = precision_score(y_test, predictions)
    recall = recall_score(y_test, predictions)
    f1 = f1_score(y_test, predictions)

    results.append({
        "Model": model_name,
        "Accuracy": accuracy,
        "Precision": precision,
        "Recall": recall,
        "F1-Score": f1
    })

results_df = pd.DataFrame(results)

results_df

,Model,Accuracy,Precision,Recall,F1-Score
0,Logistic Regression,0.803279,0.800000,0.848485,0.823529
1,KNN,0.786885,0.777778,0.848485,0.811594
2,Random Forest,0.754098,0.764706,0.787879,0.776119


In [32]:
results_df.round(4)

,Model,Accuracy,Precision,Recall,F1-Score
0,Logistic Regression,0.8033,0.8000,0.8485,0.8235
1,KNN,0.7869,0.7778,0.8485,0.8116
2,Random Forest,0.7541,0.7647,0.7879,0.7761


In [33]:
best_model_name = results_df.loc[
    results_df['F1-Score'].idxmax(),
    'Model'
]

print("Best Model based on F1-Score:", best_model_name)

Best Model based on F1-Score: Logistic Regression


## Model Evaluation

Three classification algorithms were trained: Logistic Regression, K-Nearest Neighbors, and Random Forest.

The models were evaluated using Accuracy, Precision, Recall, and F1-Score. These metrics provide a broader assessment of classification performance than accuracy alone.

The model with the highest F1-Score is selected as the best-performing model for further analysis.